File: app/src/main/java/com/app/meditec/MapsActivity.java
package com.app.meditec;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.databinding.DataBindingUtil;
import androidx.lifecycle.ViewModelProvider;
import android.annotation.SuppressLint;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.location.Location;
import android.os.Bundle;
import android.os.Looper;
import android.util.Log;
import android.view.View;
import android.widget.ImageView;
import android.widget.Toast;
import com.app.meditec.databinding.ActivityMapBinding;
import com.app.meditec.databinding.PlaceInfoBottomSheetBinding;
import com.app.meditec.models.PlaceInfo;
import com.app.meditec.ui.MapsViewModel;
import com.app.meditec.utils.PermissionUtils;
import com.google.android.gms.location.FusedLocationProviderClient;
import com.google.android.gms.location.LocationCallback;
import com.google.android.gms.location.LocationRequest;
import com.google.android.gms.location.LocationResult;
import com.google.android.gms.location.LocationServices;
import com.google.android.gms.maps.CameraUpdateFactory;
import com.google.android.gms.maps.GoogleMap;
import com.google.android.gms.maps.OnMapReadyCallback;
import com.google.android.gms.maps.SupportMapFragment;
import com.google.android.gms.maps.model.LatLng;
import com.google.android.gms.maps.model.MarkerOptions;
import com.google.android.gms.tasks.Task;
import com.google.android.libraries.places.api.Places;
import com.google.android.material.bottomsheet.BottomSheetBehavior;
import java.util.ArrayList;
import java.util.List;

public class MapsActivity extends AppCompatActivity implements OnMapReadyCallback, PermissionUtilsListener {
    private static final String TAG = "MapsActivity";
    public static final int LOCATION_PERMISSION_REQUEST_CODE = 102;
    public static final float DEFAULT_ZOOM = 15f;
    public static final int GPS_REQUEST_CODE = 189;
    private boolean mLocationPermissionGranted = false;
    private LocationCallback mLocationCallback;
    private GoogleMap mGoogleMap;
    private FusedLocationProviderClient mFusedLocationProviderClient;
    private Location mCurrentLocation;
    private static List<PlaceInfo> mPlaceInfoList;
    private BottomSheetBehavior mBottomSheetBehavior;
    private ImageView mHeaderArrow;
    private MapsViewModel mMapsViewModel;
    private ActivityMapBinding mBinding;
    private PlaceInfoBottomSheetBinding mBottomSheetBinding;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mBinding = DataBindingUtil.setContentView(this, R.layout.activity_map);
        PermissionUtils.INSTANCE.setListener(this);
        mMapsViewModel = new ViewModelProvider(this).get(MapsViewModel.class);
        mBottomSheetBinding = mBinding.bottomSheet;
        mLocationPermissionGranted = PermissionUtils.INSTANCE.getLocationPermission(this);
        if (mLocationPermissionGranted)
            initializeMap();

        createLocationCallback();
        Places.initialize(getApplicationContext(), BuildConfig.MAPS_API_KEY);
        mPlaceInfoList = new ArrayList<>();
        mBottomSheetBehavior = BottomSheetBehavior.from(mBottomSheetBinding.bottomSheet);
        mHeaderArrow = findViewById(R.id.header_arrow);
        headerImageClickListener();
        bottomSheetBehaviourCallback();
    }

    @Override
    protected void onStart() {
        super.onStart();
        mFusedLocationProviderClient = LocationServices.getFusedLocationProviderClient(this);
        if (mLocationPermissionGranted)
            PermissionUtils.INSTANCE.checkIfGPSIsEnabled(this);
    }

    @SuppressLint("MissingPermission")
    private void setupLocationUpdatesListener() {
        LocationRequest locationRequest = LocationRequest.create();
        locationRequest.setInterval(10000);
        locationRequest.setFastestInterval(10000);
        locationRequest.setPriority(LocationRequest.PRIORITY_HIGH_ACCURACY);

        mFusedLocationProviderClient.requestLocationUpdates(locationRequest, mLocationCallback, Looper.myLooper());
    }

    private void createLocationCallback() {
        mLocationCallback = new LocationCallback() {
            @Override
            public void onLocationResult(LocationResult locationResult) {
                super.onLocationResult(locationResult);
                if (locationResult != null) {
                    mCurrentLocation = locationResult.getLastLocation();
                    moveCamera(new LatLng(mCurrentLocation.getLatitude(), mCurrentLocation.getLongitude()));
                    mMapsViewModel.getPlaces(mCurrentLocation.getLatitude(), mCurrentLocation.getLongitude());
                }
            }
        };
    }

    @Override
    protected void onResume() {
        super.onResume();
        mMapsViewModel.getPlacesLiveData().observe(this, placeInfos -> {
            mPlaceInfoList = placeInfos;
            for (PlaceInfo placeInfo : mPlaceInfoList) {
                LatLng latLng = new LatLng(placeInfo.getGeometry().getLocation().getLat(),
                        placeInfo.getGeometry().getLocation().getLng());
                MarkerOptions markerOptions = new MarkerOptions()
                        .title(placeInfo.getName())
                        .position(latLng);
                mGoogleMap.addMarker(markerOptions);
            }
        });

        mMapsViewModel.getPlacesResponseStatus().observe(this, placeResponseStatus -> {
            Toast.makeText(this, placeResponseStatus, Toast.LENGTH_SHORT).show();
        });
    }

    @Override
    public void GPSIsEnabled() {
        if (mLocationPermissionGranted)
            setupLocationUpdatesListener();
    }

    @SuppressLint("MissingPermission")
    @Override
    public void onMapReady(GoogleMap googleMap) {
        Toast.makeText(this, "Map is ready", Toast.LENGTH_SHORT).show();
        mGoogleMap = googleMap;

        mGoogleMap.setMyLocationEnabled(true);
        mGoogleMap.getUiSettings().setMyLocationButtonEnabled(true);

        mGoogleMap.setOnMarkerClickListener(marker -> {
            Log.d(TAG, "Marker is clicked: " + marker.getTitle());
            Log.d(TAG, "on Marker Click: " + mPlaceInfoList.size());
            boolean isPlaceInList = getPlaceDetails(marker.getPosition());
            if (isPlaceInList) {
                if (mBottomSheetBehavior.getState() == BottomSheetBehavior.STATE_COLLAPSED)
                    mBottomSheetBehavior.setState(BottomSheetBehavior.STATE_EXPANDED);
            }
            return false;
        });

    }

    @Override
    protected void onStop() {
        super.onStop();
        if (mFusedLocationProviderClient != null)
            mFusedLocationProviderClient.removeLocationUpdates(mLocationCallback);
    }

    private boolean getPlaceDetails(LatLng position) {
        for (PlaceInfo placeInfo : mPlaceInfoList) {
            LatLng place = new LatLng(placeInfo.getGeometry().getLocation().getLat(),
                    placeInfo.getGeometry().getLocation().getLng());
            if (place.equals(position)) {
                mBottomSheetBinding.setPlaceInfo(placeInfo);
                return true;
            }
        }
        return false;
    }

    private void initializeMap() {
        SupportMapFragment mapFragment = (SupportMapFragment) getSupportFragmentManager().findFragmentById(R.id.map);
        mapFragment.getMapAsync(MapsActivity.this);
    }

    private void moveCamera(LatLng latLng) {
        mGoogleMap.animateCamera(CameraUpdateFactory.newLatLngZoom(latLng, MapsActivity.DEFAULT_ZOOM));
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == GPS_REQUEST_CODE) {
            if (resultCode == RESULT_OK) {
                if (mLocationPermissionGranted)
                    setupLocationUpdatesListener();
            } else {
                Log.d(TAG, "user ignored GPS alert");
                Toast.makeText(this, "Keep your GPS enabled", Toast.LENGTH_LONG).show();
                finish();
            }
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        Log.d(TAG, "onRequestPermissionsResult called !");
        mLocationPermissionGranted = false;

        if (requestCode == LOCATION_PERMISSION_REQUEST_CODE) {
            if (grantResults.length > 0) {
                for (int grantResult : grantResults) {
                    if (grantResult != PackageManager.PERMISSION_GRANTED) {
                        Log.d(TAG, "onRequestPermissionsResult failed");
                        Toast.makeText(this, "Location permissions needed to show maps", Toast.LENGTH_LONG).show();
                        mLocationPermissionGranted = false;
                        finish();
                        return;
                    }
                }
                Log.d(TAG, "onRequestPermissionsResult granted");
                mLocationPermissionGranted = true;
                initializeMap();
                PermissionUtils.INSTANCE.checkIfGPSIsEnabled(this);
            }
        }
    }

    /*-----------------------  bottom sheet -----------------------*/
    private void bottomSheetBehaviourCallback() {
        mBottomSheetBehavior.addBottomSheetCallback(new BottomSheetBehavior.BottomSheetCallback() {
            @Override
            public void onStateChanged(@NonNull View bottomSheet, int newState) {

            }

            @Override
            public void onSlide(@NonNull View bottomSheet, float slideOffset) {
                mHeaderArrow.setRotation(slideOffset * 180);
            }
        });
    }

    private void headerImageClickListener() {
        mHeaderArrow.setOnClickListener(v -> {
            if (mBottomSheetBehavior.getState() == BottomSheetBehavior.STATE_EXPANDED)
                mBottomSheetBehavior.setState(BottomSheetBehavior.STATE_COLLAPSED);
            else
                mBottomSheetBehavior.setState(BottomSheetBehavior.STATE_EXPANDED);
        });
    }

}

File: app/src/main/java/com/app/meditec/ui/maps/MapsActivity.kt
package com.app.meditec.ui.maps

import android.annotation.SuppressLint
import android.app.Activity
import android.content.Intent
import android.content.pm.PackageManager
import android.location.Location
import android.os.Bundle
import android.os.Looper
import android.util.Log
import android.view.View
import android.widget.AdapterView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.databinding.DataBindingUtil
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import com.app.meditec.BuildConfig
import com.app.meditec.R
import com.app.meditec.adapters.SearchPlacesAutoCompleteAdapter
import com.app.meditec.databinding.ActivityMapBinding
import com.app.meditec.databinding.DirectionsBottomSheetBinding
import com.app.meditec.databinding.PlaceInfoBottomSheetBinding
import com.app.meditec.models.PlaceInfo
import com.app.meditec.utils.MapUtils
import com.app.meditec.utils.PermissionUtils
import com.app.meditec.utils.PermissionUtilsListener
import com.google.android.gms.common.api.ApiException
import com.google.android.gms.location.*
import com.google.android.gms.maps.CameraUpdateFactory
import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.OnMapReadyCallback
import com.google.android.gms.maps.SupportMapFragment
import com.google.android.gms.maps.model.*
import com.google.android.libraries.places.api.Places
import com.google.android.libraries.places.api.model.AutocompleteSessionToken
import com.google.android.libraries.places.api.model.Place
import com.google.android.libraries.places.api.net.FetchPlaceRequest
import com.google.android.libraries.places.api.net.PlacesClient
import com.google.android.material.bottomsheet.BottomSheetBehavior
import com.google.android.material.bottomsheet.BottomSheetBehavior.BottomSheetCallback
import com.google.android.material.bottomsheet.BottomSheetBehavior.from
import java.util.*

class MapsActivity : AppCompatActivity(), OnMapReadyCallback, PermissionUtilsListener {
    private var isCameraMovedOnAppStart = false
    private var mLocationPermissionGranted = false
    private var mLocationCallback: LocationCallback? = null
    private var mGoogleMap: GoogleMap? = null
    private var mFusedLocationProviderClient: FusedLocationProviderClient? = null
    private var mCurrentLocation: Location? = null
    private var mPlaceInfoList: List<PlaceInfo>? = null
    private var mPlacesClient: PlacesClient? = null
    private var mToken: AutocompleteSessionToken? = null
    private val mPolyLines = mutableListOf<Polyline>()
    private lateinit var mBottomSheetBehavior: BottomSheetBehavior<*>
    private lateinit var mDirectionSheetBehavior: BottomSheetBehavior<*>
    private lateinit var mMapsViewModel: MapsViewModel
    private lateinit var mBinding: ActivityMapBinding
    private lateinit var mBottomSheetBinding: PlaceInfoBottomSheetBinding
    private lateinit var mDirectionsSheetBinding: DirectionsBottomSheetBinding
    private lateinit var mAutoCompleteListener: AdapterView.OnItemClickListener
    private lateinit var mSearchPlacesAutoCompleteAdapter: SearchPlacesAutoCompleteAdapter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        mBinding = DataBindingUtil.setContentView(this, R.layout.activity_map)
        PermissionUtils.setListener(this)
        mMapsViewModel = ViewModelProvider(this).get(MapsViewModel::class.java)
        mBottomSheetBinding = mBinding.bottomSheet
        mDirectionsSheetBinding = mBinding.directionsSheet
        mLocationPermissionGranted = PermissionUtils.getLocationPermission(this)
        if (mLocationPermissionGranted) initializeMap()
        createLocationCallback()
        Places.initialize(applicationContext, BuildConfig.MAPS_API_KEY)
        mPlacesClient = Places.createClient(this)
        mToken = AutocompleteSessionToken.newInstance()
        mSearchPlacesAutoCompleteAdapter = SearchPlacesAutoCompleteAdapter(this, mPlacesClient!!, mToken!!)
        mPlaceInfoList = ArrayList()
        mBottomSheetBehavior = from(mBottomSheetBinding.bottomSheet)
        mDirectionSheetBehavior = from(mDirectionsSheetBinding.directionsSheet)
        initViews()
        bottomSheetCallback()
    }

    override fun onStart() {
        super.onStart()
        mFusedLocationProviderClient = LocationServices.getFusedLocationProviderClient(this)
        if (mLocationPermissionGranted) PermissionUtils.checkIfGPSIsEnabled(this)
    }

    @SuppressLint("MissingPermission")
    private fun setupLocationUpdatesListener() {
        val locationRequest = LocationRequest.create()
        locationRequest.interval = 10000
        locationRequest.fastestInterval = 10000
        locationRequest.priority = LocationRequest.PRIORITY_HIGH_ACCURACY
        mFusedLocationProviderClient!!.requestLocationUpdates(locationRequest, mLocationCallback, Looper.myLooper())
    }

    private fun createLocationCallback() {
        mLocationCallback = object : LocationCallback() {
            override fun onLocationResult(locationResult: LocationResult?) {
                super.onLocationResult(locationResult)
                if (locationResult != null) {
                    mCurrentLocation = locationResult.lastLocation
                    mMapsViewModel.userCurrentLocation = locationResult.lastLocation
                    if (!isCameraMovedOnAppStart) {
                        isCameraMovedOnAppStart = true
                        moveCamera(LatLng(mCurrentLocation!!.latitude, mCurrentLocation!!.longitude))
                    }
                }
            }
        }
    }

    override fun onResume() {
        super.onResume()
        mMapsViewModel.placesLiveData.observe(this, Observer { placeInfos ->
            mPlaceInfoList = placeInfos
            for ((name, _, _, _, geometry) in mPlaceInfoList!!) {
                val latLng = LatLng(geometry.location.lat, geometry.location.lng)
                mGoogleMap!!.addMarker(MapUtils.createMarkerOptions(name, latLng))
            }
        })
        mMapsViewModel.routeLiveData.observe(this, Observer { routes ->
            mBottomSheetBehavior.state = BottomSheetBehavior.STATE_COLLAPSED
            for (polyline in mPolyLines) polyline.remove()
            mPolyLines.clear()

            for (route in routes) {
                val polylineOptions = PolylineOptions()
                polylineOptions.color(resources.getColor(R.color.colorAccent))
                polylineOptions.width(7f)
                polylineOptions.addAll(route.polyLines)
                mPolyLines.add(mGoogleMap!!.addPolyline(polylineOptions))
            }
            mDirectionSheetBehavior.state = BottomSheetBehavior.STATE_EXPANDED
            mDirectionsSheetBinding.route = routes[0]
        })
        mMapsViewModel.responseStatus.observe(this, Observer { placeResponseStatus ->
            showToast(placeResponseStatus)
        })
    }

    override fun GPSIsEnabled() {
        if (mLocationPermissionGranted) setupLocationUpdatesListener()
    }

    @SuppressLint("MissingPermission")
    override fun onMapReady(googleMap: GoogleMap) {
        showToast("Map is ready")
        mGoogleMap = googleMap
        mGoogleMap!!.isMyLocationEnabled = true
        mGoogleMap!!.uiSettings.isMyLocationButtonEnabled = true
        mGoogleMap!!.setPadding(0, 150, 0, 0)
        mGoogleMap!!.setOnMarkerClickListener { marker: Marker ->
            showPlaceDetails(marker.position)
            false
        }
    }

    override fun onStop() {
        super.onStop()
        if (mFusedLocationProviderClient != null) mFusedLocationProviderClient!!.removeLocationUpdates(mLocationCallback)
    }

    override fun onBackPressed() {
        when {
            mBottomSheetBehavior.state == BottomSheetBehavior.STATE_EXPANDED ->
                mBottomSheetBehavior.state = BottomSheetBehavior.STATE_COLLAPSED
            mDirectionSheetBehavior.state == BottomSheetBehavior.STATE_EXPANDED ->
                mDirectionSheetBehavior.state = BottomSheetBehavior.STATE_COLLAPSED
            else ->
                super.onBackPressed()
        }
    }

    private fun showPlaceDetails(position: LatLng) {
        for (placeInfo in mPlaceInfoList!!) {
            val place = LatLng(placeInfo.geometry.location.lat,
                    placeInfo.geometry.location.lng)
            if (place == position) {
                mBottomSheetBinding.placeInfo = placeInfo
                mMapsViewModel.currentlySelectedPlace = placeInfo
                if (mBottomSheetBehavior.state == BottomSheetBehavior.STATE_COLLAPSED)
                    mBottomSheetBehavior.state = BottomSheetBehavior.STATE_EXPANDED
                break
            }
        }
    }

    private fun initializeMap() {
        val mapFragment = supportFragmentManager.findFragmentById(R.id.map) as SupportMapFragment?
        mapFragment!!.getMapAsync(this)
    }

    private fun moveCamera(latLng: LatLng) {
        mGoogleMap!!.animateCamera(CameraUpdateFactory.newLatLngZoom(latLng, DEFAULT_ZOOM))
    }

    private fun moveCameraAndAddMarker(latLng: LatLng, title: String) {
        moveCamera(latLng)
        mGoogleMap!!.addMarker(MapUtils.createMarkerOptions(title, latLng))
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == GPS_REQUEST_CODE) {
            if (resultCode == Activity.RESULT_OK) {
                if (mLocationPermissionGranted) setupLocationUpdatesListener()
            } else {
                Log.d(TAG, "user ignored GPS alert")
                showToast("Keep your GPS enabled")
                finish()
            }
        }
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        mLocationPermissionGranted = false
        if (requestCode == LOCATION_PERMISSION_REQUEST_CODE) {
            if (grantResults.isNotEmpty()) {
                for (grantResult in grantResults) {
                    if (grantResult != PackageManager.PERMISSION_GRANTED) {
                        showToast("Location permissions needed to show maps")
                        mLocationPermissionGranted = false
                        finish()
                        return
                    }
                }
                mLocationPermissionGranted = true
                initializeMap()
                PermissionUtils.checkIfGPSIsEnabled(this)
            }
        }
    }

    private fun initViews() {
        mBinding.autoCompleteTextView.setAdapter(mSearchPlacesAutoCompleteAdapter)

        mAutoCompleteListener = AdapterView.OnItemClickListener { _, _, position, _ ->
            val item = mSearchPlacesAutoCompleteAdapter.getItem(position)
            val placeId = item?.placeId
            getPlaceWith(placeId)
        }

        mBinding.autoCompleteTextView.onItemClickListener = mAutoCompleteListener

        mDirectionsSheetBinding.closeImageButton.setOnClickListener {
            mDirectionSheetBehavior.state = BottomSheetBehavior.STATE_COLLAPSED
        }

        mBottomSheetBinding.detailsFab.setOnClickListener { mMapsViewModel.onClickGoFab() }
        mDirectionsSheetBinding.busImageButton.setOnClickListener { mMapsViewModel.onTransitImageButton() }
        mDirectionsSheetBinding.carImageButton.setOnClickListener { mMapsViewModel.onDriveImageButton() }
        mDirectionsSheetBinding.walkImageButton.setOnClickListener { mMapsViewModel.onWalkImageButton() }

    }

    private fun getPlaceWith(placeId: String?) {
        val placeFields = listOf(Place.Field.ID, Place.Field.NAME, Place.Field.LAT_LNG)
        val request = FetchPlaceRequest.builder(placeId!!, placeFields).build()

        mPlacesClient!!.fetchPlace(request).addOnSuccessListener { response ->
            val place = response.place
            moveCameraAndAddMarker(place.latLng!!, place.name!!)
        }.addOnFailureListener { exception ->
            if (exception is ApiException) {
                showToast("An error occurred, place not found")
            }
        }
    }

    private fun bottomSheetCallback() {
        mBottomSheetBehavior.addBottomSheetCallback(object : BottomSheetCallback() {
            override fun onSlide(bottomSheet: View, slideOffset: Float) {}
            override fun onStateChanged(bottomSheet: View, newState: Int) {
                if (newState != BottomSheetBehavior.STATE_COLLAPSED)
                    mGoogleMap!!.setPadding(0, 150, 0, bottomSheet.height)
                else
                    mGoogleMap!!.setPadding(0, 150, 0, 0)
            }
        })
        mDirectionSheetBehavior.addBottomSheetCallback(object : BottomSheetCallback() {
            override fun onSlide(bottomSheet: View, slideOffset: Float) {}
            override fun onStateChanged(bottomSheet: View, newState: Int) {
                if (newState == BottomSheetBehavior.STATE_COLLAPSED)
                    mBinding.autoCompleteTextView.visibility = View.VISIBLE
                else
                    mBinding.autoCompleteTextView.visibility = View.INVISIBLE
            }
        })
    }

    private fun showToast(message: String) {
        Toast.makeText(this, message, Toast.LENGTH_SHORT).show()
    }

    companion object {
        private const val TAG = "MapsActivity"
        const val LOCATION_PERMISSION_REQUEST_CODE = 102
        const val DEFAULT_ZOOM = 15f
        const val GPS_REQUEST_CODE = 189
    }
}